pubnub/models/subscription.py

```diff
@@ -0,0 +1,413 @@
+            return subscription.endswith('.*') and name.startswith(subscription.strip('*'))
+
+        if self._type == PNSubscriptionType.CHANNEL:
+            return message.channel == self.name or wildcard_match(message.channel, self.name)
+        else:
+            return message.subscription == self.name
+
+    def presence(self, presence):
+        if not hasattr(self, 'on_presence') or not self.with_presence:
+            return
+
+        if self.is_matching_listener(presence) and hasattr(self, 'on_presence'):
+            self.on_presence(presence)
+
+    def message(self, message):
+        if self.is_matching_listener(message) and hasattr(self, 'on_message'):
+            self.on_message(message)
+
+    def message_action(self, message_action):
+        if self.is_matching_listener(message_action) and hasattr(self, 'on_message_action'):
+            self.on_message_action(message_action)
+
+    def signal(self, signal):
+        if self.is_matching_listener(signal) and hasattr(self, 'on_signal'):
+            self.on_signal(signal)
+
+
+class PNSubscribeCapable:
+    def subscribe(self, timetoken: Optional[int] = None, region: Optional[str] = None,
+                  with_presence: Optional[bool] = None):
+        self.timetoken = timetoken
+        self.region = region
+        self.with_presence = with_presence
+        self.subscription_registry.add(self)
+
+    def unsubscribe(self):
+        self.subscription_registry.remove(self)
+
+
+class PubNubSubscription(PNEventEmitter, PNSubscribeCapable):
+    def __init__(self, pubnub_instance, name: str, type: PNSubscriptionType, with_presence: bool = False) -> None:
+        self.subscription_registry = pubnub_instance._subscription_registry
+        self.subscription_manager = pubnub_instance._subscription_manager
+        self.name = name
+        self._type = type
+        self.with_presence = with_presence
+
+    def add_listener(self, listener):
+        self.subscription_registry.add_listener(listener)
+
+    def get_names_with_presence(self):
+        return [self.name, f'{self.name}-pnpres'] if self.with_presence else [self.name]
+
+
+class PubNubSubscriptionSet(PNEventEmitter, PNSubscribeCapable):
+    __on_message: callable
+    __on_signal: callable
+    __on_presence: callable
+    __on_channel_metadata: callable
+    __on_user_metadata: callable
+    __on_message_action: callable
+    __on_membership: callable
+    __on_file: callable
+
+    def __init__(self, pubnub_instance, subscriptions: List[PubNubSubscription]) -> None:
+        self.subscription_registry = pubnub_instance._subscription_registry
+        self.subscription_manager = pubnub_instance._subscription_manager
+        self.subscriptions = subscriptions
+
+    @property
+    def on_message(self):
+        return self.__on_message
+
+    @on_message.setter
+    def on_message(self, callback):
+        self.__on_message = callback
+        for subscription in self.subscriptions:
```

Comment:
does that mean that if I have two subscriptions in the SubscriptionSet that match the same message, I will get the message twice from the SubscriptionSet? according to spec it should be emitted only once (i.e. if any subscription in the set matches the message, emit the message once)